Small fixes and more tests for jdk Converters

lrytz opened the pull request:
Include tests from java8-compat, and add some new ones.

Fixes https://github.com/scala/bug/issues/11468

lrytz committed:
Rough scalcheck tests for Accumulators

lrytz committed:
Port StreamConvertersTest from java8-compat

lrytz committed:
Port StepperConversionTest from java8-compat

lrytz committed:
Test for Steppers

There are a few open issues
  - ChampStepper doesn't return all elements (Rex is working on a fix)
  - TableStepper.trySplit has a bug, it was changed to always return
    `null` in a previous commit (Rex is also fixing this)
  - The `ORDERED` flag is often not correct, in particular when creating
    a Stepper of a Set, and there's no specific implementation. Then
    the implementation falls back to IterableOnce, and IteratorStepper
    has the `ORDERED` flag

lrytz commented:
Also has a few fixes
  - Adds `baseStepperShape`, needed because `anyStreamShapePrototype` actually can work on arbitrary Steppers, not just `AnyStepper`
  - Add `asJavaSeqStream` / `asJavaParStream` to Byte/Short/Char/Float arrays
  - `WrappedString.stepper` works correctly when requesting an `AnySepper` instead of an `IntStepper`...

Review by @Ichoran

Ichoran committed:
11468 - Fixed coverage in Champ trySplit and hang in Table trySplit

All `trySplit` instances now pass in collection-laws.

The Champ split is not as even as it could be; better would be
to extract the sizes from the nodes and use that to wisely pick
the partition points.  However, it's not binary compatible.

lrytz committed:
Remove workarounds

Ichoran commented:
I'll try to get to this tomorrow--I won't have time today.  Thanks for grabbing my fixes.

Ichoran reviewed (approved):
Since it's mostly tests I didn't read through exceedingly carefully; everything looks good that I saw.  The one thing I caught is that the newest collections don't seem to be part of the tests: CollisionProofHashMap, VectorMap, and TreeSeqMap.  But collection-laws tests those three and covers many of the cases here, so I don't think it's critical.

(I'm a little wary of the mima exceptions because they might mask actual incompatibilities later if we need to change something, but I guess we're starting bincompat with RC1?)

lrytz committed:
Don't use TableStepper for LinkedHashMap/Set

lrytz committed:
clean up StepperTest, clarify ORDERED flag

lrytz committed:
add CollisionProofHashMap to StepperTest

lrytz commented:
> CollisionProofHashMap, VectorMap, and TreeSeqMap

added those

> wary of the mima exceptions

We'll reset MiMa to 0 on the 2.13.0 release, so we can still break bin compat for now.

Ichoran commented:
Changes look good.  I was uncertain at the time whether to use `TableStepper` for `LinkedHashMap`, but I think you're right that the correct decision is to not.

lrytz commented:
Thanks!

lrytz commented:
@adriaanm this is ready.

lrytz committed:
Add missing shape classes for boxed Java primitive types

lrytz committed:
un-comment some tests that work now

lrytz commented:
I added a few more commits. The last one is to be discussed. I changed the references to primitives to the boxed types in the explicit conversion methods. For example

```diff
-    def toJavaOptionalDouble(o: Option[Double]): OptionalDouble = o match {
+    def toJavaOptionalDouble(o: Option[jl.Double]): OptionalDouble = o match {
```

The reason is that Scala primitives in generic signatures erase to `Object`, so the `toJavaOptionalDouble` is really not useful for Java programmers - even though that's the primary audience. For Scala programmers, it's recommended to use the extension methods in `OptionConverters.Ops._` instead.

The problem is that using the explicit conversion methods in Scala is now pretty annoying, which might be very confusing to poeple.

As an alternative, we could duplicate the `XConverters`, have a version for Java and one for Scala. I'm not sure what's the best way forward....

smarter commented:
> The reason is that Scala primitives in generic signatures erase to Object

Do you know why?

lrytz commented:
https://github.com/scala/bug/issues/4214 / https://github.com/scala/scala/commit/e42733e9fe1f3af591976fbb48b66035253d85b9

lrytz commented:
@adriaanm I removed the last commit (to use boxed types instead of scala primitives) here, will split it off to a separte PR. So this one is ready.

lrytz committed:
Port FutureConvertersTest from java8-compat

lrytz commented:
@viktorklang I copied the `FutureConvertersTest` over from java8-compat, and there are failures... Could you take a look? I'm not sure how to find out what's going on.

viktorklang commented:
@lrytz I think I fixed the test here: https://github.com/viktorklang/scala/commit/c141ee3b4a7161a458074d788388fe5ede144ef9

The problem is that the callback is both executed synchronously and completed synchronously, which means that the calling thread is waiting for the latch to trigger, so it cannot trigger the latch. I.e. the test is wrongly written.

I "fixed" the test by changing the order of the completion and the countdown, but an equal solution would be to switch to *Async versions of the methods (sadly there are no such options for all of the methods tested.

I also fixed a test which demands that obtrude throws UnsupportedOperationException.

Let me know if you want to discuss this further!

viktorklang commented:
One item of note that I discovered while doing this is that there does not seem to be a cohesive story about whether to call converters toX or asX

Ichoran commented:
It is supposed to be the case that `asX` means that it's a view onto the underlying whatever, whereas `toX` dispenses with the original after the new thing is constructed.

Maybe this implementation detail is not sufficiently important to warrant a seemingly random selection of `to` vs `as` methods, or maybe the rule isn't consistently applied, but that's what it's _supposed_ to be.

lrytz commented:
I followed that rule consistently in the `scala.jdk` package, hopefully I got it right everywhere.

lrytz commented:
@viktorklang will look tomorrow at your proposed changes, thanks. I wonder why the test passes on java8-compat (with 2.13.0-RC1, https://github.com/scala/scala-java8-compat/pull/139).

viktorklang commented:
@lrytz Hmmm, I do not know why it succeeds on the other version. Let me know if you have any follow-up questions tomorrow.

lrytz commented:
Oh, I see that I skipped a lot of method overrides when porting [FuturesConvertersImpl.CF](https://github.com/scala/scala-java8-compat/blob/master/src/main/scala/scala/concurrent/java8/FutureConvertersImpl.scala#L27) over to the [Scala repository](https://github.com/scala/scala/blob/2.13.x/src/library/scala/concurrent/impl/FutureConvertersImpl.scala#L23). Adding them fixes the test. So this is totally my fault, sorry...

lrytz committed:
Add missing overrides to FutureConvertersImpl.CF

Not sure why I missed those initially...

viktorklang commented:
@lrytz Np! I'm glad the test caught it though! :-)

He-Pin commented on src/library/scala/concurrent/impl/FutureConvertersImpl.scala:
```diff
@@ -12,19 +12,84 @@
+
+    override def thenCompose[U](fn: JFunction[_ >: T, _ <: CompletionStage[U]]): CompletableFuture[U] = thenComposeAsync(fn)
+
+    override def whenComplete(fn: BiConsumer[_ >: T, _ >: Throwable]): CompletableFuture[T] = whenCompleteAsync(fn)
```
@lrytz @SethTisue @viktorklang Why all these methods are delegated to the async ones?

lrytz replied:
There is a comment above:

```
    // Ensure that completions of this future cannot hold the Scala Future's completer hostage
```

It seems these overrides were always there https://github.com/scala/scala-java8-compat/commit/b3fca14e9ffe920b0b1ebb9f9f0f104c0869b6fc

Haven't thought about it in detail, but maybe there are separate thread pools for Scala and Java futures, and it's to avoid running on the wrong one?

som-snytt replied:
Amazingly, I could scroll back three weeks in Discord and find the link I posted.

https://github.com/scala/scala-java8-compat/blob/main/src/main/scala/scala/compat/java8/FutureConverters.scala#L57
